app/utility/base_knowledge_svc.py

```diff
@@ -0,0 +1,318 @@
+        """
+        raise NotImplementedError
+
+    def _locate(self, object_name, query=None):
+        """
+        Locate a matching object in the internal store
+        :param object_name: object type
+        :param query: dictionary of fields to match on
+        :return: list of matching objects
+        """
+        try:
+            return [obj for obj in self.fact_ram[object_name] if obj.match(query)]
+        except Exception as e:
+            self.log.error('[!] LOCATE: %s' % e)
+
+    def _remove(self, object_name, query):
+        """
+        Remove objects from the internal store
+        :param object_name: object type
+        :param query: dictionary of fields to match on
+        """
+        try:
+            sublist = self.fact_ram[object_name] = [obj for obj in self.fact_ram[object_name] if obj.match(query)]
+            self._clear_matching_constraints(sublist)
+            self.fact_ram[object_name][:] = [obj for obj in self.fact_ram[object_name] if obj not in sublist]
+        except Exception as e:
+            self.log.error('[!] REMOVE: %s' % e)
+
+    def _clear_matching_constraints(self, objs):
+        """
+        Remove constraints associated with objects as part of deletion
+        :param objs: list of objects being removed
+        """
+        need_to_remove = self._get_matching_constraints(objs)
+        for obj in need_to_remove:
+            del self.fact_ram['constraints'][obj._knowledge_id]
+
+    def _get_matching_constraints(self, objs):
+        """
+        Identify matching contstraings associated with objects
+        :param objs: list of objects to get constraints for
+        :return: list of relevant constraints
+        """
+        return [obj for obj in objs if obj._knowledge_id in self.fact_ram['constraints']]
+
+    @staticmethod
+    def _delete_file(path):
+        if not os.path.exists(path):
+            return
+        elif os.path.isdir(path):
+            shutil.rmtree(path)
+        else:
+            os.remove(path)
+
+    @staticmethod
+    def _destroy():
+        """Reset the caldera data directory and server state.
+
+        This creates a gzipped tarball backup of the data files tracked by caldera.
+        Paths are preserved within the tarball, with all files having "data/" as the
+        root.
+        """
+        if not os.path.exists(DATA_BACKUP_DIR):
+            os.mkdir(DATA_BACKUP_DIR)
+
+        timestamp = datetime.utcnow().strftime('%Y%m%d%H%M%S')
+        tarball_path = os.path.join(DATA_BACKUP_DIR, f'backup-{timestamp}.tar.gz')
+
+        with tarfile.open(tarball_path, 'w:gz') as tarball:
+            tarball.add(FACT_STORE_PATH)
+            BaseKnowledgeService._delete_file(FACT_STORE_PATH)
+
+    async def _save_state(self):
+        """
+        Save the current internal store state
+        """
+        await self.get_service('file_svc').save_file('object_store', pickle.dumps(self.fact_ram), 'data')
```

Comment:
Is `object_store` descriptive enough as file name? (not sure what the other data/object files are named)

Reply:
Yeah, I can get behind re-naming this so that it's more descriptive.